### PR TITLE
Implement adding empty public partitions

### LINKED ISSUE #110: Adding missing public partitions
# Context

Atm public partition implementation is not complete. Filtering out partitions which are not in `public_partitions` is performed ([code](https://github.com/OpenMined/PipelineDP/blob/545166054b8ac1750dafa752165323b08ac7ee8e/pipeline_dp/dp_engine.py#L63)), but adding those public partitions which are not in data has not been implemented. This issue for implementing that.

# Design

Method [DPEngine.aggregate](https://github.com/OpenMined/PipelineDP/blob/545166054b8ac1750dafa752165323b08ac7ee8e/pipeline_dp/dp_engine.py#L43) does high-level management of the computation graph (contribution bounding, partition selection etc). The computation graph now is the following.

1. Drop non public partitions ([code](https://github.com/OpenMined/PipelineDP/blob/545166054b8ac1750dafa752165323b08ac7ee8e/pipeline_dp/dp_engine.py#L63))
2. Bound contributions (and aggregation per (**partition_key**, **privacy_id**))
3. Aggregation per **partition_key**.
4. Computing DP metrics.

 It looks like a pretty good place (please advice on better places if there are) to add missing public partition is between steps 2 and 3. We can merge (flatten) existed collection of (partition_key, acumulator) with (**public_partitions**, **accumulator_for_empty_set**). 

[aggregate_fn](https://github.com/OpenMined/PipelineDP/blob/545166054b8ac1750dafa752165323b08ac7ee8e/pipeline_dp/dp_engine.py#L60) should return **accumulator_for_empty_set** when called with **[]**.

### PR DESCRIPTION
**Pull request by miracvbasaran**
## Description
Fixes #110.

## Affected Dependencies
n/a

## How has this been tested?
- Unit tests
- movie ratings example

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [x] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [x] My changes are covered by tests

### CODE REVIEW

File: tests/dp_engine_test.py [hunk 1]
```diff
@@ -400,7 +400,7 @@ def test_aggregate_private_partition_selection_drop_many(self):
         # Most partition should be dropped by private partition selection.
         # This tests is non-deterministic, but it should pass with probability
         # very close to 1.
-        self.assertLess(len(col), 5)
+        self.assertLess(5, len(col))
```
Comment (miracvbasaran) [inline tests/dp_engine_test.py:403]: "expected" first, then "actual"

File: pipeline_dp/dp_engine.py [hunk 2]
```diff
@@ -180,6 +187,19 @@ def _drop_not_public_partitions(self, col, public_partitions,
             f"Public partition selection: dropped non public partitions")
         return self._ops.map_tuple(col, lambda k, v: v, "Drop key")
 
+    def _add_empty_public_partitions(self, col, public_partitions,
+                                     aggregator_fn):
+        self._add_report_stage(
+            "Adding empty accumulators to public partitions and joining "
```
Comment (dvadym) [inline pipeline_dp/dp_engine.py:193]: Those reports are for API users. Accumulators are not part of the API. 
Maybe something like that
"Adding empty partitions for those public partitions which are missing in data" 

Technically speaking right here we're doing in slightly different way, but it's equavalent. WDYT?

File: tests/dp_engine_test.py (hunk 1 above)
Comment (dvadym) [reply tests/dp_engine_test.py:403]: Thanks!

File: pipeline_dp/dp_engine.py [hunk 3]
```diff
@@ -180,6 +187,19 @@ def _drop_not_public_partitions(self, col, public_partitions,
             f"Public partition selection: dropped non public partitions")
         return self._ops.map_tuple(col, lambda k, v: v, "Drop key")
 
+    def _add_empty_public_partitions(self, col, public_partitions,
+                                     aggregator_fn):
```
Comment (dvadym) [inline pipeline_dp/dp_engine.py:191]: Please add description to this function

**Review by dvadym — COMMENTED** [review]
Thanks, it looks great! I left minor comments

File: pipeline_dp/pipeline_operations.py [hunk 4]
```diff
@@ -86,6 +86,14 @@ def reduce_accumulators_per_key(self, col, stage_name: str):
         """
         pass
 
+    @abc.abstractmethod
+    def flatten(self, col1, col2, stage_name: str):
```
Comment (dvadym) [inline pipeline_dp/pipeline_operations.py:90]: The problem is that there are other PipelineOperations, and w/o implementaiton this method in them, they are abstract clasess which can't be created.

Please add this method in other PipelineOperations:

In MultiProcessPipelineOperations please use implementation from LocalPipelineOpeations

In SparkPipelineOperations please add this method with implementation
`raise NotImplementedException("...") `
I'll implement it
In Spark please

**Review by dvadym — APPROVED** [review]

**Pull request merged**